File: day4/day4.py
#!/usr/bin/env python3

import re
import sys

VALIDATORS_PART1 = (
    ("byr", lambda x: True),
    ("iyr", lambda x: True),
    ("eyr", lambda x: True),
    ("hgt", lambda x: True),
    ("hcl", lambda x: True),
    ("ecl", lambda x: True),
    ("pid", lambda x: True),
)

yr_pat = re.compile(r"^\d{4}$")
hgt_pat = re.compile(r"^(\d+)(in|cm)$")
hcl_pat = re.compile(r"^#[0-9a-f]{6}$")
ecl_pat = re.compile(r"^(amb|blu|brn|gry|grn|hzl|oth)$")
pid_pat = re.compile(r"^\d{9}$")

def hgt_check(x):
    m = hgt_pat.match(x)
    if m is None:
        return False
    elif m[2] == "cm":
        return 150 <= int(m[1]) <= 193
    elif m[2] == "in":
        return 59 <= int(m[1]) <= 76
    else:
        return False

VALIDATORS_PART2 = (
    ("byr", lambda x: yr_pat.match(x) and 1920 <= int(x) <= 2002),
    ("iyr", lambda x: yr_pat.match(x) and 2010 <= int(x) <= 2020),
    ("eyr", lambda x: yr_pat.match(x) and 2020 <= int(x) <= 2030),
    ("hgt", hgt_check),
    ("hcl", lambda x: bool(hcl_pat.match(x))),
    ("ecl", lambda x: bool(ecl_pat.match(x))),
    ("pid", lambda x: bool(pid_pat.match(x))),
)

def is_valid(passport, validators):
    for field, fn in validators:
        value = passport.get(field)
        if not (value and fn(value)):
            return False

    return True

def process_batch(txt, validators):
    count = 0
    passport = {}
    for line in txt.split("\n"):
        line = line.strip()
        if line == "":
            if is_valid(passport, validators):
                count += 1

            passport = {}
        else:
            for token in line.split():
                field, value = token.split(":")
                passport[field] = value

    if is_valid(passport, validators):
        count += 1

    return count

# VALIDATORS_MAP = {x:y for x, y in VALIDATORS_PART2}
# assert VALIDATORS_MAP["byr"]("2002") == True
# assert VALIDATORS_MAP["byr"]("2003") == False
# assert VALIDATORS_MAP["hgt"]("60in") == True
# assert VALIDATORS_MAP["hgt"]("190cm") == True
# assert VALIDATORS_MAP["hgt"]("190in") == False
# assert VALIDATORS_MAP["hgt"]("190") == False
# assert VALIDATORS_MAP["hcl"]("#123abc") == True
# assert VALIDATORS_MAP["hcl"]("#123abz") == False
# assert VALIDATORS_MAP["hcl"]("123abc") == False
# assert VALIDATORS_MAP["ecl"]("brn") == True
# assert VALIDATORS_MAP["ecl"]("wat") == False
# assert VALIDATORS_MAP["pid"]("000000001") == True
# assert VALIDATORS_MAP["pid"]("0123456789") == False

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "r") as f:
            txt = f.read()
    except:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)

    print("Part1:", process_batch(txt, VALIDATORS_PART1))
    print("Part2:", process_batch(txt, VALIDATORS_PART2))

File: day4/day4.cpp
#include <fstream>
#include <functional>
#include <regex>
#include <unordered_map>
#include <vector>
#include <fmt/format.h>

using namespace std;

using Validator = pair<string, function<bool(const string&)>>;

const vector<Validator> VPART1 = {
	{ "byr", [](const string&){ return true; } },
	{ "iyr", [](const string&){ return true; } },
	{ "eyr", [](const string&){ return true; } },
	{ "hgt", [](const string&){ return true; } },
	{ "hcl", [](const string&){ return true; } },
	{ "ecl", [](const string&){ return true; } },
	{ "pid", [](const string&){ return true; } },
};

const regex pass_pat(R"(^(.*):(.*)$)");
const regex year_pat(R"(^\d{4}$)");
const regex hgt_pat(R"(^(\d+)(in|cm)$)");
const regex hcl_pat(R"(^#[0-9a-f]{6}$)");
const regex ecl_pat(R"(^(amb|blu|brn|gry|grn|hzl|oth)$)");
const regex pid_pat(R"(^\d{9}$)");

const vector<Validator> VPART2 = {
	{
		"byr",
		[](const string& s) {
			if (!regex_match(s, year_pat))
			{
				return false;
			}
			int value = stoi(s);
			return 1920 <= value && value <= 2002;
		}
	},
	{
		"iyr",
		[](const string& s) {
			if (!regex_match(s, year_pat))
			{
				return false;
			}
			int value = stoi(s);
			return 2010 <= value && value <= 2020;
		}
	},
	{
		"eyr",
		[](const string& s) {
			if (!regex_match(s, year_pat))
			{
				return false;
			}
			int value = stoi(s);
			return 2020 <= value && value <= 2030;
		}
	},
	{
		"hgt",
		[](const string& s) {
			smatch sm;
			if (!regex_match(s, sm, hgt_pat))
			{
				return false;
			}
			int value = stoi(sm[1]);
			if (sm[2] == "cm")
			{
				return 150 <= value && value <= 193;
			}
			else
			{
				return 59 <= value && value <= 76;
			}
		}
	},
	{
		"hcl",
		[](const string& s) {
			return regex_match(s, hcl_pat);
		}
	},
	{
		"ecl",
		[](const string& s) {
			return regex_match(s, ecl_pat);
		}
	},
	{
		"pid",
		[](const string& s) {
			return regex_match(s, pid_pat);
		}
	},
};

struct Passport
{
	unordered_map<string,string> fields;

	void clear()
	{
		fields.clear();
	}

	void add_field(const string& record)
	{
		smatch sm;
		if (regex_match(record, sm, pass_pat))
		{
			fields[sm[1]] = sm[2];
		}
	}

	bool is_valid(const vector<Validator>& validators) const
	{
		for (auto& v: validators)
		{
			auto it = fields.find(v.first);
			if (!(it != fields.end() && v.second(it->second)))
			{
				return false;
			}
		}
		return true;
	}
};

istream& operator>>(istream& in, Passport& pass)
{
	bool read = false;
	pass.clear();
	string line;
	while (getline(in, line) && line.size() != 0)
	{
		size_t pos = 0;
		size_t next;
		while ((next = line.find(' ', pos)) != string::npos)
		{
			pass.add_field(line.substr(pos, next - pos));
			pos = next + 1;
		}
		pass.add_field(line.substr(pos));
		read = true;
	}
	// NOTE: hack to allow to read the last element without
	// discarding it because of the EOF signalling
	if (read) in.clear();
	return in;
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		fmt::print(stderr, "Usage: {} <filename>\n", argv[0]);
		return 1;
	}

	ifstream input(argv[1]);
	if (!input)
	{
		fmt::print(stderr, "Cannot open {}", argv[1]);
		return 1;
	}

	int part1 = 0;
	int part2 = 0;
	Passport pass;
	while (input >> pass)
	{
		part1 += pass.is_valid(VPART1);
		part2 += pass.is_valid(VPART2);
	}
	input.close();

	fmt::print("Part1: {}\nPart2: {}\n", part1, part2);
	return 0;
}
